src/cmd/keys.js

```diff
@@ -21,208 +24,154 @@ const dfu = require('../lib/dfu');
+			const protocol = await this.validateDeviceProtocol({ device });
+			let alg = this._getPrivateKeyAlgorithm({ protocol });
+			let prefilename = path.join(path.dirname(filename), 'backup_' + alg + '_' + path.basename(filename));
+			await this._saveKeyFromDevice({ filename: prefilename, force: true, device });
+			let segmentName = this._getPrivateKeySegmentName({ protocol });
+			let segment = this._validateSegmentSpecs(segmentName);
+			const buffer = fs.readFileSync(filename, null); // 'null' to get the raw data
+			await device.writeOverDfu(buffer, { altSetting: segment.specs.alt, startAddr: segment.specs.address, size: segment.specs.size, noErase: true, leave: leave });
+
+			console.log('Key written to device!');
+		} catch (err) {
+			throw new VError(ensureError(err), 'Error writing key to device.');
+		} finally {
+			if (device) {
+				await device.close();
+			}
+		}
 	}
 
-	saveKeyFromDevice({ force, params: { filename } }){
+	async saveKeyFromDevice({ force, params: { filename } }){
 		filename = utilities.filenameNoExt(filename) + '.der';
-		return this._saveKeyFromDevice({ filename, force });
+		await this._saveKeyFromDevice({ filename, force });
 	}
 
-	_saveKeyFromDevice({ filename, force }){
-		const { tryDelete, filenameNoExt, deferredChildProcess } = utilities;
+	async _saveKeyFromDevice({ filename, force, device }) {
 		let protocol;
+		const { tryDelete, filenameNoExt, deferredChildProcess } = utilities;
 
-		if (!force && fs.existsSync(filename)){
+		if (!force && fs.existsSync(filename)) {
 			throw new VError('This file already exists, please specify a different file, or use the --force flag.');
-		} else if (fs.existsSync(filename)){
+		} else if (fs.existsSync(filename)) {
 			tryDelete(filename);
 		}
 
-		//find this.dfu devices, make sure a device is connected
-		//pull the key down and save it there
+		const keepDeviceOpen = device ? true : false;
 
-		return Promise.resolve()
-			.then(() => this.dfu.isDfuUtilInstalled())
-			.then(() => this.dfu.findCompatibleDFU())
-			.then(() => this.validateDeviceProtocol())
-			.then(_protocol => {
-				protocol = _protocol;
-				let segment = this._getPrivateKeySegmentName({ protocol });
-				return this.dfu.read(filename, segment, false);
-			})
-			.then(() => {
-				let pubPemFilename = filenameNoExt(filename) + '.pub.pem';
+		try {
+			if (!device) {
+				device = await this.getDfuDevice();
+			}
+			protocol = await this.validateDeviceProtocol({ device });
+			let segmentName = this._getPrivateKeySegmentName({ protocol });
+			let segment = this._validateSegmentSpecs(segmentName);
+			let buf;
+			try {
+				buf = await device.readOverDfu({ altSetting: segment.specs.alt, startAddr: segment.specs.address, size: segment.specs.size });
+			} catch (err) {
+				// FIXME: First time read may fail so we retry
+				buf = await device.readOverDfu({ altSetting: segment.specs.alt, startAddr: segment.specs.address, size: segment.specs.size });
+			}
```

Comment:
As mentioned, there is a challenge when calling `writeOverDfu` and it is followed by a `readOverDfu`. I noticed that there is a USB Stall Error during the first read operation which gets cleared during the second read operation. I am currently debugging this but added a workaround here to temporarily unblock this issue.